src/main/java/io/quarkus/gizmo/Type.java

```diff
@@ -0,0 +1,418 @@
+package io.quarkus.gizmo;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.PrimitiveType.Primitive;
+
+/**
+ * This interface can be used to build a JVM signature for classes, methods and fields.
+ * <p>
+ * Implementations are instantiated via factory methods; for example {@link #voidType()} and {@link #classType(Class)}.
+ * 
+ * @see SignatureBuilder
+ */
+public interface Type {
+
+    // factory methods
+
+    public static VoidType voidType() {
+        return VoidType.INSTANCE;
+    }
+
+    public static PrimitiveType byteType() {
+        return PrimitiveType.BYTE;
+    }
+
+    public static PrimitiveType booleanType() {
+        return PrimitiveType.BOOLEAN;
+    }
+
+    public static PrimitiveType intType() {
+        return PrimitiveType.INT;
+    }
+
+    public static PrimitiveType longType() {
+        return PrimitiveType.LONG;
+    }
+
+    public static PrimitiveType shortType() {
+        return PrimitiveType.SHORT;
+    }
+
+    public static PrimitiveType doubleType() {
+        return PrimitiveType.DOUBLE;
+    }
+
+    public static PrimitiveType floatType() {
+        return PrimitiveType.FLOAT;
+    }
+
+    public static PrimitiveType charType() {
+        return PrimitiveType.CHAR;
+    }
+
+    public static ClassType classType(DotName name) {
+        return new ClassType(Objects.requireNonNull(name).toString().replace('.', '/'));
+    }
+
+    public static ClassType classType(String name) {
+        return new ClassType(name);
+    }
+
+    public static ClassType classType(Class<?> classType) {
+        Objects.requireNonNull(classType);
+        if (classType.getDeclaringClass() != null) {
```

Comment:
What does this do? Is it some special case for nested classes? If so, it only works for one level of nesting, no?

Reply:
Yes, nested classes are a bit problematic. And yes, you're right that this is wrong ;-)

Reply:
The use case we don't support at the moment is a nested class declared on parameterized type.

Reply:
I've removed this logic.